_posts/2020-07-12-the-lifecycle-of-spring-beans.md

```diff
@@ -0,0 +1,427 @@
+    System.out.println("--- init-method executed ---");
+  }
+
+  @PreDestroy
+  public void preDestroy() {
+    System.out.println("--- @PreDestroy executed ---");
+  }
+
+  @Override
+  public void destroy() throws Exception {
+    System.out.println("--- destroy executed ---");
+  }
+
+  public void destroyMethod() {
+    System.out.println("--- destroy-method executed ---");
+  }
+
+}
+```
+Besides, we create a `BeanPostProcessor` to hook into the before and after initialization phases:
+```java
+public class MyBeanPostProcessor implements BeanPostProcessor {
+
+  @Override
+  public Object postProcessBeforeInitialization(Object bean, String beanName)
+      throws BeansException {
+    if (bean instanceof MySpringBean) {
+      System.out.println("--- postProcessBeforeInitialization executed ---");
+    }
+    return bean;
+  }
+
+  @Override
+  public Object postProcessAfterInitialization(Object bean, String beanName)
+      throws BeansException {
+    if (bean instanceof MySpringBean) {
+      System.out.println("--- postProcessAfterInitialization executed ---");
+    }
+    return bean;
+  }
+
+}
+```
+Next, we write a Spring configuration to define our beans:
+```java
+@Configuration
+public class MySpringConfiguration {
+
+  @Bean
+  public MyBeanPostProcessor myBeanPostProcessor(){
+    return new MyBeanPostProcessor();
+  }
+
+  @Bean(initMethod = "initMethod", destroyMethod = "destroyMethod")
+  public MySpringBean mySpringBean(){
+    return new MySpringBean();
+  }
+
+}
+```
+Finally, we write a `@SpringBootTest` to run our Spring context:
+```java
+@SpringBootTest
+public class BeanLifecycleApplicationTests {
+
+  @Autowired
+  public MySpringBean mySpringBean;
+
+  @Test
+  public void testMySpringBeanLifecycle() {
+    String message = "Hello World";
+    mySpringBean.sendMessage(message);
+    assertThat(mySpringBean.getMessage()).isEqualTo(message);
+  }
+
+}
+```
+
+As a result, our test method outputs the execution order between the lifecycle phases:
+```java
+--- setBeanName executed ---
+--- setApplicationContext executed ---
+--- postProcessBeforeInitialization executed ---
+--- @PostConstruct executed ---
+--- afterPropertiesSet executed ---
+--- init-method executed ---
+--- postProcessAfterInitialization executed ---
+...
+--- @PreDestroy executed ---
+--- destroy executed ---
+--- destroy-method executed ---
+```
+
+## Conclusion
+
+In this tutorial, we explained what the bean lifecycle phases are, why, and how we hook into lifecycle phases in Spring.
+
+Spring has numerous phases in a bean lifecycle as well as many ways to receive callbacks. We can hook into these phases both over on beans or from a common point as we do in `BeanPostProcessor`.
+
+Although each method has its purpose, **we should avoid using Spring interfaces. This will force us to couple our code to Spring Framework and violates the <a href="https://en.wikipedia.org/wiki/Dependency_inversion_principle">Dependency Inversion Principle.</a>**
```

Comment:
Why does it violate the DIP? Can you elaborate in a sentence?

Reply:
I think I referred to the DIP because it is a way of decoupling the components, modules, etc. But now another thought, it was a little too much to say it directly violates the principle itself. So, I think I'm gonna remove it and only address that we should decouple our code as much as we can even from the Spring Framework if possible.